src/runner.c

```diff
@@ -0,0 +1,134 @@
+  fse_t *instance = (fse_t *)data;
+
+  uv_mutex_lock(&instance->mutex);
+  for (idx=0; idx < numEvents; idx++) {
+    fse_event_t *event = (fse_event_t*)malloc(sizeof(fse_event_t));
+    CFStringRef path = (CFStringRef)CFArrayGetValueAtIndex((CFArrayRef)eventPaths, idx);
+    event->pathlen = CFStringGetLength(path);
+    strncpy(event->path, CFStringGetCStringPtr(path, kCFStringEncodingUTF8), PATH_MAX);
+    event->id = eventIds[idx];
+    event->flags = eventFlags[idx];
+    event->next = NULL;
+    if (instance->head) {
+      ((fse_event_t*)instance->head)->next = event;
+      instance->head = event;
+    } else {
+      instance->head = event;
+      instance->tail = event;
+    }
+  }
+  uv_async_send(&instance->async);
+  uv_mutex_unlock(&instance->mutex);
+}
+
+void eventCallback(uv_async_t *async) {
+  fse_t *instance = (fse_t *) async->data;
+  napi_handle_scope scope;
+  napi_value this, callback, argv[3], result;
+  fse_event_t *event;
+  uv_mutex_lock(&instance->mutex);
+  assert(napi_open_handle_scope(instance->env, &scope) == napi_ok);
+  assert(napi_get_null(instance->env, &this) == napi_ok);
+  assert(napi_get_reference_value(instance->env, instance->callback, &callback) == napi_ok);
+  instance->head = NULL;
+  while (instance->tail) {
+    event = (fse_event_t*)instance->tail;
+    instance->tail = event->next;
+    assert(napi_create_string_utf8(instance->env, event->path, event->pathlen, &argv[0]) == napi_ok);
+    assert(napi_create_uint32(instance->env, event->flags, &argv[1]) == napi_ok);
+    assert(napi_create_int64(instance->env, event->id, &argv[2]) == napi_ok);
+    napi_call_function(instance->env, this, callback, 3, argv, &result);
+    free(event);
+  }
+  assert(napi_close_handle_scope(instance->env, scope) == napi_ok);
+  uv_mutex_unlock(&instance->mutex);
+}
+
+void execute(void *data) {
+  fse_t *instance = (fse_t*)data;
+
+  instance->loop = CFRunLoopGetCurrent();
+
+  FSEventStreamContext context = { 0, data, NULL, NULL, NULL };
+  CFStringRef dirs[] = { CFStringCreateWithCString(NULL, instance->path, kCFStringEncodingUTF8) };
+  FSEventStreamRef stream = FSEventStreamCreate(NULL, &handleEvents, &context, CFArrayCreate(NULL, (const void **)&dirs, 1, NULL), kFSEventStreamEventIdSinceNow, (CFAbsoluteTime) 0.1, kFSEventStreamCreateFlagNone | kFSEventStreamCreateFlagWatchRoot | kFSEventStreamCreateFlagFileEvents | kFSEventStreamCreateFlagUseCFTypes);
+  FSEventStreamScheduleWithRunLoop(stream, instance->loop, kCFRunLoopDefaultMode);
+  FSEventStreamStart(stream);
+
+  CFRunLoopRun();
+
+  FSEventStreamStop(stream);
+  FSEventStreamUnscheduleFromRunLoop(stream, instance->loop, kCFRunLoopDefaultMode);
+  FSEventStreamInvalidate(stream);
+  FSEventStreamRelease(stream);
+
+  instance->loop = NULL;
+}
+
+void cleanup(napi_env env, void* data, void* ignored) {
+  fse_t *instance = (fse_t*)data;
+  stop(instance);
+  assert(napi_delete_reference(instance->env, instance->callback) == napi_ok);
+  free(instance);
```

Comment:
This is a use-after-free; you can't free the memory until the `uv_close()` callback is called.

Reply:
???
Why
???
At this point the thread is dead. Line#101 joins the thread. After this nothing could possibly use this. How would `uv_close()` come into it / how would I listen that to `free(instance)`?

Reply:
`fse_t` embeds the `uv_async_t` and libuv isn't done with that until it invokes the close callback. Options you have are:

1. Pass a callback that calls `free(handle->data)` (obviously don't NULL it first), or
2. Make the `uv_async_t` the first member of `fse_t` and pass `(uv_close_cb) free` to `uv_close()`, or
3. Change `uv_async_t async` to `uv_async *async` so the handle can outlive the `fse_t` struct.

Which one is best is a matter of personal taste. Option 2 is a little more efficient - less pointer indirection - but that's insignificant here.

If you do pick 2, add a comment to the struct explaining that `async` should be the first member.

Reply:
I am going with 3, because stopping / freeing of instance may not happen at the same time, so I need to free them at distinct points from each other.